linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/TextWriter.scala

```diff
@@ -0,0 +1,611 @@
+    )
+
+    localNames = None
+    labelNames = None
+    labelNameGen = None
+  }
+
+  private def writeTag(tag: Tag)(implicit b: WatBuilder): Unit = {
+    b.newLineList(
+      "tag", {
+        appendName(tag.id)
+        writeTypeUse(tag.typeID)
+      }
+    )
+  }
+
+  private def writeGlobal(g: Global)(implicit b: WatBuilder) =
+    b.newLineList(
+      "global", {
+        appendName(g.id)
+        if (g.isMutable)
+          b.sameLineList("mut", writeType(g.tpe))
+        else writeType(g.tpe)
+        g.init.instr.foreach(writeInstr)
+      }
+    )
+
+  private def writeExport(e: Export)(implicit
+      b: WatBuilder
+  ) = b.newLineList(
+    "export", {
+      b.appendElement("\"" + e.name + "\"")
+      e.desc match {
+        case ExportDesc.Func(id) =>
+          b.sameLineList(
+            "func",
+            { appendName(id) }
+          )
+        case ExportDesc.Global(id) =>
+          b.sameLineList(
+            "global",
+            { appendName(id) }
+          )
+      }
+    }
+  )
+
+  private def writeStart(startFunction: FunctionID)(implicit b: WatBuilder): Unit = {
+    b.newLineList(
+      "start", {
+        appendName(startFunction)
+      }
+    )
+  }
+
+  private def writeElement(element: Element)(implicit b: WatBuilder): Unit = {
+    b.newLineList(
+      "elem", {
+        element.mode match {
+          case Element.Mode.Declarative => b.appendElement("declare")
+        }
+        writeType(element.tpe)
+        element.init.foreach { item =>
+          b.newLineList(
+            "item",
+            item.instr.foreach(writeInstr(_))
+          )
+        }
+      }
+    )
+  }
+
+  private def writeData(data: Data)(implicit b: WatBuilder): Unit = {
+    b.newLineList(
+      "data", {
+        appendName(data.id)
+        data.mode match {
+          case Data.Mode.Passive => ()
+        }
+        b.appendElement("\"" + data.bytes.map("\\%02x".format(_)).mkString + "\"")
+      }
+    )
+  }
+
+  private def writeTypeUse(typeID: TypeID)(implicit b: WatBuilder): Unit = {
+    b.sameLineList("type", appendName(typeID))
+  }
+
+  private def writeType(tpe: StorageType)(implicit b: WatBuilder): Unit = {
+    tpe match {
+      case tpe: SimpleType => b.appendElement(tpe.textName)
+      case tpe: PackedType => b.appendElement(tpe.textName)
+
+      case RefType(true, heapType: HeapType.AbsHeapType) =>
+        b.appendElement(heapType.nullableRefTextName)
+
+      case RefType(nullable, heapType) =>
+        b.sameLineList(
+          "ref", {
+            if (nullable)
+              b.appendElement("null")
+            writeHeapType(heapType)
+          }
+        )
+    }
+  }
+
+  private def writeHeapType(heapType: HeapType)(implicit b: WatBuilder): Unit = {
+    heapType match {
+      case HeapType.Type(typeID)          => appendName(typeID)
+      case heapType: HeapType.AbsHeapType => b.appendElement(heapType.textName)
+    }
+  }
+
+  private def floatString(v: Double): String = {
```

Comment:
`writeFloatString`? This is only ever used to actually write the string. Returning it feels very inconsistent.